diff --git a/init/main.c b/init/main.c
index 004c75de5..e39d11e56 100644
--- a/init/main.c
+++ b/init/main.c
@@ -1069,6 +1069,10 @@ static int __ref kernel_init(void *unused)
 
 	rcu_end_inkernel_boot();
 
+	pr_info(" ");
+	pr_info("#### Johnny Li (UFID: 9699-4228) How I figured this out is a miracle. #####");
+	pr_info(" ");
+
 	if (ramdisk_execute_command) {
 		ret = run_init_process(ramdisk_execute_command);
 		if (!ret)
